### PR TITLE
[4.0] Kill grunt!

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue # .

### Summary of Changes
Two years ago (or so) I did this Grunt file with the aim to lift some of the day to day tasks for maintainers and also provide an easy way for contributors to compile scss files and minify js and css.
As the project (J4) moved on we added a lot more functionality and the file reached about 800 lines of code, which is kinda awkward for a task runner file.
Also we already encounter (a few times) that plugins will fall out of sync with the latest releases of the underlaying scripts (postfix, uglify, etc).
Therefore I propose a a better (for the long term) solution: plain node js code, no Grunt, no Gulp, no Webpack or Rollup. The tasks are pretty straight forward and we're not doing any packaging so webpack and rollup are totally out of the picture.
Also we are now using the package.json as the only point of truth (used to be another yaml file)

### Available Commands
- Compile css
```
node build --compilecss
```

- Prepare the installation language.js
```
node build --installer
```

- Compile/compress js
```
node build --compilejs
```

- Update all vendor dependencies
```
node build --update
```

### Expected result
Same functionality as Grunt (some command)

### Documentation Changes Required
For documentation just run 
```
node build --help
```
or just
```
node build
```
Enjoy!

### Preview
screen shot 2017-10-30 at 13 19 11

### CODE REVIEW

**Comment by dgrammatiko** [comment]
I forgot to mention that:
compilecss and compilejs options produce ALWAYS the same output, compared to the mess we had in grunt!

**Comment by C-Lodder** [comment]
You appear to be loading the wrong jQuery UI package from NPM.

Should be this: https://www.npmjs.com/package/jquery-ui
not this: https://www.npmjs.com/package/jquery-ui-dist

**Comment by C-Lodder** [comment]
Also, can we kill the uncompressed CSS files please? We're not using them at all.

File: package.json [hunk 1]
```diff
@@ -2,78 +2,297 @@
...
         "babelify",
         {
           "presets": [
-            "es2015"
+            [
+              "env",
+              {
+                "targets": {
+                  "browsers": [
+                    "last 2 versions",
```
Comment (C-Lodder) [inline package.json]: should be `last 1 version`

File: build/build-modules-js/update.js [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+			concatFiles([
+				"media/vendor/codemirror/addon/display/fullscreen.css",
+				"media/vendor/codemirror/addon/fold/foldgutter.css",
+				"media/vendor/codemirror/addon/scroll/simplescrollbars.css"
+			], 'media/vendor/codemirror/lib/addons.css');
+
+			// Update the XML files for Codemirror
+			let codemirrorXml = fs.readFileSync(rootPath + '/plugins/editors/codemirror/codemirror.xml', {encoding: 'UTF-8'});
+			codemirrorXml = codemirrorXml.replace(xmlVersionStr, "$1" + options.dependencies.tinymce + "$3");
```
Comment (C-Lodder) [inline build/build-modules-js/update.js]: `options.dependencies.tinymce` should be `options.dependencies.codemirror` as you've reverted Codemirror back to 4.7.1 :)

**Comment by mbabker** [comment]
> Also, can we kill the uncompressed CSS files please? We're not using them at all.

Unless you're going to kill support for the core templates and core media loading uncompressed files in debug mode they should stay.

**Comment by C-Lodder** [comment]
@mbabker They don't exist now and the templates work perfectly fine in debug mode

**Comment by C-Lodder** [comment]
@mbabker - Atum doesn't have any uncompressed CSS files and works perfectly fine in debug mode: https://github.com/joomla/joomla-cms/tree/4.0-dev/administrator/templates/atum/css

**Comment by mbabker** [comment]
I know it all works fine in debug mode.  JHtml only uses uncompressed if it can take a `.min.foo` filename, convert it to `.foo`, and find a file that exists.  So if uncompressed media doesn't exist then the compressed file still gets loaded.

This might be OK for the template media, but it should still be possible to get the core media in uncompressed format.  Which means either we keep them in repo & ship them or people have to go through extra steps to compile them.

File: package.json (hunk 1 above)
Comment (dgrammatiko) [reply package.json]: this one is for babel, not postcss

File: build/build-modules-js/update.js (hunk 2 above)
Comment (dgrammatiko) [reply build/build-modules-js/update.js]: oops

**Comment by Fedik** [comment]
@dgt41 it is still WIP?

I got error:
screen 2017-11-02 17 55 32 970x275

**Comment by Fedik** [comment]
@dgt41 ignore me, all works fine

**Comment by Fedik** [comment]
@dgt41 one note:
```
       "srcjs": "dist",
       "srccss": "dist",
        "filesjs": [
          "cropper.js",
          "cropper.min.js",
          "cropper.js.map"
        ],
```

I found it confusing, because you have `"srcjs": "dist",` and `dist` usually contain `distributive` files, not `source` 😉 

And question.
Is there no way to get the module path, and scan for available files, automatically?
something like :
```
var pathToModule = require.resolve('moduleBlabLa');
var files = scanforJsCss(pathToModule);
```

**Comment by dgrammatiko** [comment]
@Fedik that's an interesting idea, I have to explore that

**Comment by C-Lodder** [comment]
Could we get the conflicts on this fixed please and then merge?

Any improvements can be made at a later date

**Comment by dgrammatiko** [comment]
Grrrr 
screen shot 2017-12-31 at 11 24 01

**Comment by C-Lodder** [comment]
Just recompile and remove the grunt-settings.yaml

**Comment by dgrammatiko** [comment]
@wilsonge **important** please merge this!!

**Comment by C-Lodder** [comment]
@wilsonge please merge. This is a choice made by the JS team and will also assist the Ciaran and myself who are the primary people making use of the feature.

**Comment by dgrammatiko** [comment]
@wilsonge any decision here, before you merge more things and make this harder to get it again in sync?

**Comment by wilsonge** [comment]
Fix the conflicts and I'll merge it. I'm very uneasy about this to be honest. I still think gulp might be better for us. But let's just get it in so it's not blocking you guys

**Comment by dgrammatiko** [comment]
This is in-sync again. Ping @wilsonge

**Comment by wilsonge** [comment]
You need to fix travis please

**Comment by dgrammatiko** [comment]
Travis is happy again

**Comment by dgrammatiko** [comment]
Thank you!

**Comment by infograf768** [comment]
@dgt41 
I'm afraid this has broken, among other stuff, the radio field on Firefox.
See https://github.com/joomla/joomla-cms/issues/19403

**Comment by C-Lodder** [comment]
We are in the process pf fixing things

**Pull request merged**